sdk/tests/utilities/test_token_refresh.py

```diff
@@ -186,3 +194,388 @@ def force_refresh(refresh_token):
+                json_data={
+                    "access_token": "mock_access_token_2",
+                    "refresh_token": "mock_refresh_token",
+                    "expires_in": 60
+                },
+                status_code=200
+            ),
+        ]):
+            # Ensure that we were able to get the first access token
+            self.assertEqual(f"{refreshing_token}", "mock_access_token")
+
+            sleep(1)  # Wait for initial token to expire
+
+            # Try and get access token again forcing refresh, if we can get it then retry was called
+            self.assertEqual(f"{refreshing_token}", "mock_access_token_2")
+
+    def test_does_not_retry_on_4xx_status_code_other_than_429(self):
+        """
+        Ensures that we do not retry on other common 4xx status codes such as 400 - Bad Request
+        """
+        refreshing_token = RefreshingToken(api_configuration=self.config)
+
+        with patch("requests.post", side_effect=[
+            # Return a 429 on the first attempt
+            MockApiResponse(
+                json_data={
+                    "error": "rate_limit",
+                    "error_description": "API rate limit exceeded."
+                },
+                status_code=429,
+                headers={}
+            ),
+            # Return a 400 on the second attempt
+            MockApiResponse(
+                json_data={
+                    "error": "invalid_grant",
+                    "error_description": "The refresh token is invalid or expired."
+                },
+                status_code=400
+            ),
+        ]):
+
+            # Ensure that a 400 is raised as an error and not retried
+            with self.assertRaises(ValueError):
+                self.force_refresh(refreshing_token)
+
+    def test_retries_on_429s_up_till_retry_limit(self):
+        """
+        Ensures that the refreshing token only retries up until the retry limit to prevent
+        an infinite retry loop
+        """
+        refreshing_token = RefreshingToken(api_configuration=self.config)
+
+        refreshing_token.retry_limit = 2  # Override default to ensure test runs in reasonable amount of time
+        expected_requests = 1 + refreshing_token.retry_limit  # Initial request plus expected number retries
+
+        with patch("requests.post") as identity_mock:
+            identity_mock.side_effect = [
+                MockApiResponse(
+                    json_data={
+                        "error": "rate_limit",
+                        "error_description": "API rate limit exceeded."
+                    },
+                    status_code=429,
+                    headers={}
+                )
+            ] * expected_requests  # Return a 429 every time up until expected number of attempts
+
+            # Ensure that a an error is raised once reaching the retry limit
+            with self.assertRaises(ValueError):
+                self.force_refresh(refreshing_token)
```

Comment:
is it possible to assert on the specific error?

Reply:
Added

`self.assertIn("Max retry limit", str(retry_limit_error.exception))`